fix(core): implement proper token counting for mixed content in getNumTokens

christian-bromann opened the pull request:
<!--
Thank you for contributing to LangChain.js! Your PR will appear in our next release under the title you set above. Please make sure it highlights your valuable contribution.

To help streamline the review process, please make sure you read our contribution guidelines:
https://github.com/langchain-ai/langchainjs/blob/main/CONTRIBUTING.md

If you are adding an integration (e.g. a new LLM, vector store, or memory), please also read our additional guidelines for integrations:
https://github.com/langchain-ai/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

Replace this block with a description of the change, the issue it fixes (if applicable), and relevant context.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle below!
-->

## Description

This PR fixes a critical issue in the `BaseLanguageModel.getNumTokens()` method where it would return 0 tokens for any non-string `MessageContent`, effectively ignoring text content in mixed-content messages.

### Problem
Previously, when calling `getNumTokensFromMessages()` with messages containing mixed content (e.g., text + image_url arrays), the text portions were not being counted, leading to inaccurate token usage calculations. The method had a TODO comment and simply returned 0 for any array content:

```typescript
// TODO: Figure out correct value.
if (typeof content !== "string") {
  return 0;
}
```

### Solution
Implemented proper text extraction logic that:
- **Maintains backward compatibility**: String content works exactly as before
- **Extracts text from arrays**: Processes `MessageContentComplex[]` to find and concatenate all text blocks
- **Ignores non-text content**: Image URLs, audio, and other media content are properly ignored for token counting
- **Handles edge cases**: Empty arrays, image-only content, and multiple text blocks

### Example
```javascript
const messages = [
  new HumanMessage({
    content: [
      {
        type: "text",
        text: "What is this image?", // This text is now properly counted
      },
      {
        type: "image_url",
        image_url: {
          url: "https://example.com/image.jpg",
        },
      },
    ],
  }),
];

// Before: getNumTokensFromMessages would undercount due to 0 tokens from getNumTokens
// After: Correctly counts tokens for "What is this image?"
```

### Changes Made
- **Core fix**: Replaced TODO implementation in `getNumTokens()` with proper text extraction
- **Comprehensive tests**: Added test suite covering mixed content, edge cases, and compatibility
- **Test restructuring**: Improved test organization with `describe/it` blocks for better clarity

### Testing
- ✅ All existing tests pass (backward compatibility maintained)
- ✅ New tests cover mixed content scenarios
- ✅ Edge cases tested (empty arrays, image-only content, multiple text blocks)
- ✅ Integration with `getNumTokensFromMessages` verified

This fix ensures accurate token counting for modern chat applications that use rich message content with text, images, and other media types.

Fixes #8310

hntrl reviewed (changes requested):
Better than what we had before but I don't know if this is "end-state" token counting (for instance in python we consider other things like function calls apart of the token count, there's probably other ones we should be considering). Probably at a minimum we re-add the comment to signal a more appropriate solution is needed. It would be interesting to compare different message content instances with gpt-2 tokenizer and this fallback method.

I'll also call out I don't think the blast radius on this is big (I think) since we only use this in the trimMessages utility

https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/messages/utils.py#L86

christian-bromann commented:
Thanks for reviewing!

> I'll also call out I don't think the blast radius on this is big (I think) since we only use this in the trimMessages utility

There have been 2 reported instances of this failing for users, see #8310 and #8336. I agree with you, this is definitely not perfect but addresses the issues reported. I've added a comment to revisit this section at a later time. Let me know if there is anything else I can do.
